File: examples/language-impl-comparison/ocaml5/mini_lisp.ml
(* ミニ Lisp 評価機 (OCaml 5 実装)
   S式構文を持つ式を解析して評価する
   OCaml 5 の代数的効果を活用 *)

module MiniLisp = struct
  (* 式の抽象構文木 *)
  type expr =
    | Number of float
    | Symbol of string
    | List of expr list

  (* 評価値 *)
  type value =
    | VNumber of float
    | VLambda of {
        params : string list;
        body : expr;
        env : env;
      }
    | VBuiltin of (value list -> (value, string) result)

  and env = (string * value) list

  (* パースエラー *)
  type parse_error =
    | UnexpectedToken of string
    | UnmatchedParen
    | EmptyInput

  (* トークン化: S式の括弧をスペースで区切る *)
  let tokenize source =
    source
    |> String.split_on_char ' '
    |> List.concat_map (fun s ->
        s
        |> String.split_on_char '('
        |> List.concat_map (fun s2 -> String.split_on_char ')' s2)
      )
    |> List.filter (fun s -> s <> "")

  (* 式のパース *)
  let rec parse_expr tokens =
    match tokens with
    | [] -> Error EmptyInput
    | token :: rest -> parse_token token rest

  and parse_token token rest =
    if token = "(" then
      parse_list rest []
    else if token = ")" then
      Error UnmatchedParen
    else
      match Float.of_string_opt token with
      | Some num -> Ok (Number num, rest)
      | None -> Ok (Symbol token, rest)

  and parse_list tokens acc =
    match tokens with
    | [] -> Error UnmatchedParen
    | ")" :: rest -> Ok (List (List.rev acc), rest)
    | token :: rest ->
      (match parse_token token rest with
       | Ok (expr, next) -> parse_list next (expr :: acc)
       | Error err -> Error err)

  (* 式の評価 *)
  let rec eval_expr expr env =
    match expr with
    | Number n -> Ok (VNumber n)
    | Symbol name ->
      (match List.assoc_opt name env with
       | Some value -> Ok value
       | None -> Error ("未定義シンボル: " ^ name))
    | List items -> eval_list items env

  and eval_list items env =
    match items with
    | [] -> Error "空のリストは評価できません"
    | head :: rest ->
      (match eval_expr head env with
       | Error err -> Error err
       | Ok callee ->
         (match evaluate_args rest env with
          | Error err -> Error err
          | Ok args -> apply callee args))

  and evaluate_args exprs env =
    List.fold_left
      (fun acc_res expr ->
         match acc_res with
         | Error err -> Error err
         | Ok acc ->
           (match eval_expr expr env with
            | Ok value -> Ok (acc @ [value])
            | Error err -> Error err))
      (Ok [])
      exprs

  and apply callee args =
    match callee with
    | VBuiltin fn -> fn args
    | VLambda { params; body; env } -> apply_lambda params body env args
    | VNumber _ -> Error "数値を関数として適用できません"

  and apply_lambda params body lambda_env args =
    if List.length params <> List.length args then
      Error "引数の数が一致しません"
    else
      let new_env = List.combine params args @ lambda_env in
      eval_expr body new_env

  (* 組み込み数値演算 *)
  let builtin_numeric op =
    fun args ->
      match args with
      | [VNumber lhs; VNumber rhs] -> Ok (VNumber (op lhs rhs))
      | _ -> Error "数値演算は2引数の数値のみ対応します"

  (* デフォルト環境 *)
  let default_env () =
    [
      ("+", VBuiltin (builtin_numeric ( +. )));
      ("-", VBuiltin (builtin_numeric ( -. )));
      ("*", VBuiltin (builtin_numeric ( *. )));
      ("/", VBuiltin (builtin_numeric ( /. )));
    ]

  (* メイン評価関数 *)
  let eval source =
    let tokens = tokenize source in
    match parse_expr tokens with
    | Error EmptyInput -> Error "入力が空です"
    | Error UnmatchedParen -> Error "括弧が一致しません"
    | Error (UnexpectedToken token) -> Error ("予期しないトークン: " ^ token)
    | Ok (expr, rest) ->
      if rest = [] then
        eval_expr expr (default_env ())
      else
        Error "末尾に未消費トークンがあります"
end

(* 利用例 *)
(* MiniLisp.eval "(+ 40 2)" => Ok (VNumber 42.0) *)

File: examples/language-impl-comparison/scala3/mini_lisp.scala
// ミニ Lisp 評価機 (Scala 3 実装)
// S式構文を持つ式を解析して評価する

package miniLisp

import scala.util.{Try, Success, Failure}

// 式の抽象構文木
enum Expr:
  case Number(value: Double)
  case Symbol(name: String)
  case List(items: scala.collection.immutable.List[Expr])

// 評価値
enum Value:
  case VNumber(value: Double)
  case VLambda(params: scala.collection.immutable.List[String], body: Expr, env: Env)
  case VBuiltin(fn: scala.collection.immutable.List[Value] => Either[String, Value])

type Env = Map[String, Value]

// パースエラー
enum ParseError:
  case UnexpectedToken(token: String)
  case UnmatchedParen
  case EmptyInput

// トークン化: S式の括弧をスペースで区切る
def tokenize(source: String): scala.collection.immutable.List[String] =
  source
    .replace("(", " ( ")
    .replace(")", " ) ")
    .split("\\s+")
    .toList
    .filter(_.nonEmpty)

// 式のパース
def parseExpr(tokens: scala.collection.immutable.List[String]): Either[ParseError, (Expr, scala.collection.immutable.List[String])] =
  tokens match
    case Nil => Left(ParseError.EmptyInput)
    case token :: rest => parseToken(token, rest)

def parseToken(token: String, rest: scala.collection.immutable.List[String]): Either[ParseError, (Expr, scala.collection.immutable.List[String])] =
  if token == "(" then
    parseList(rest, Nil)
  else if token == ")" then
    Left(ParseError.UnmatchedParen)
  else
    Try(token.toDouble) match
      case Success(num) => Right((Expr.Number(num), rest))
      case Failure(_) => Right((Expr.Symbol(token), rest))

def parseList(tokens: scala.collection.immutable.List[String], acc: scala.collection.immutable.List[Expr]): Either[ParseError, (Expr, scala.collection.immutable.List[String])] =
  tokens match
    case Nil => Left(ParseError.UnmatchedParen)
    case ")" :: rest => Right((Expr.List(acc.reverse), rest))
    case token :: rest =>
      parseToken(token, rest) match
        case Right((expr, next)) => parseList(next, expr :: acc)
        case Left(err) => Left(err)

// 式の評価
def evalExpr(expr: Expr, env: Env): Either[String, Value] =
  expr match
    case Expr.Number(n) => Right(Value.VNumber(n))
    case Expr.Symbol(name) =>
      env.get(name) match
        case Some(value) => Right(value)
        case None => Left(s"未定義シンボル: $name")
    case Expr.List(items) => evalList(items, env)

def evalList(items: scala.collection.immutable.List[Expr], env: Env): Either[String, Value] =
  items match
    case Nil => Left("空のリストは評価できません")
    case head :: rest =>
      for
        callee <- evalExpr(head, env)
        args <- evaluateArgs(rest, env)
        result <- apply(callee, args)
      yield result

def evaluateArgs(exprs: scala.collection.immutable.List[Expr], env: Env): Either[String, scala.collection.immutable.List[Value]] =
  exprs.foldLeft[Either[String, scala.collection.immutable.List[Value]]](Right(Nil)) { (acc, expr) =>
    for
      values <- acc
      value <- evalExpr(expr, env)
    yield values :+ value
  }

def apply(callee: Value, args: scala.collection.immutable.List[Value]): Either[String, Value] =
  callee match
    case Value.VBuiltin(fn) => fn(args)
    case Value.VLambda(params, body, lambdaEnv) => applyLambda(params, body, lambdaEnv, args)
    case Value.VNumber(_) => Left("数値を関数として適用できません")

def applyLambda(params: scala.collection.immutable.List[String], body: Expr, lambdaEnv: Env, args: scala.collection.immutable.List[Value]): Either[String, Value] =
  if params.length != args.length then
    Left("引数の数が一致しません")
  else
    val newEnv = lambdaEnv ++ params.zip(args).toMap
    evalExpr(body, newEnv)

// 組み込み数値演算
def builtinNumeric(op: (Double, Double) => Double): scala.collection.immutable.List[Value] => Either[String, Value] =
  args => args match
    case scala.collection.immutable.List(Value.VNumber(lhs), Value.VNumber(rhs)) => Right(Value.VNumber(op(lhs, rhs)))
    case _ => Left("数値演算は2引数の数値のみ対応します")

// デフォルト環境
def defaultEnv(): Env = Map(
  "+" -> Value.VBuiltin(builtinNumeric(_ + _)),
  "-" -> Value.VBuiltin(builtinNumeric(_ - _)),
  "*" -> Value.VBuiltin(builtinNumeric(_ * _)),
  "/" -> Value.VBuiltin(builtinNumeric(_ / _))
)

// メイン評価関数
def eval(source: String): Either[String, Value] =
  val tokens = tokenize(source)
  parseExpr(tokens) match
    case Left(ParseError.EmptyInput) => Left("入力が空です")
    case Left(ParseError.UnmatchedParen) => Left("括弧が一致しません")
    case Left(ParseError.UnexpectedToken(token)) => Left(s"予期しないトークン: $token")
    case Right((expr, rest)) =>
      if rest.isEmpty then
        evalExpr(expr, defaultEnv())
      else
        Left("末尾に未消費トークンがあります")

// 利用例
// eval("(+ 40 2)") => Right(VNumber(42.0))